### PR TITLE
Optimize validator

### PR DESCRIPTION
**Pull request by murtukov**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

Change the way the validation rules are generated. As result the generated classes and the `TypeBuilder` class reduced in size.

#### Before

Before the change all validation rules were genereted **inside** the resolver closure and passed to the `InputValidator`:

```php
public function __construct(ConfigProcessor $configProcessor, GraphQLServices $services)
{
    $config = [
        'name' => self::NAME,
        'fields' => fn() => [
            'linkedConstraintsValidation' => [
                'type' => Type::boolean(),
                'resolve' => function ($value, $args, $context, $info) use ($services) {
                    $validator = new InputValidator(
                        \func_get_args(),
                        $services->get('container')->get('validator'),
                        $services->get('validatorFactory'),
                        [
                            'string1' => [
                                'link' => ['App\Models\User', 'string1', 'property'],
                            ],
                            'string2' => [
                                'link' => ['App\Models\User', 'string2', 'getter'],
                            ],
                            'string3' => [
                                'link' => ['App\Models\User', 'string3', 'member'],
                            ],
                        ],
                    );
                    return $services->mutation("mutation_mock", $args, $validator);
                },
                'args' => [
                    [
                        'name' => 'string1',
                        'type' => Type::nonNull(Type::string()),
                    ],
                    [
                        'name' => 'string2',
                        'type' => Type::nonNull(Type::string()),
                    ],
                    [
                        'name' => 'string3',
                        'type' => Type::nonNull(Type::string()),
                    ],
                ],
            ],
            'cascadeValidationWithGroups' => [
                'type' => Type::boolean(),
                'resolve' => function ($value, $args, $context, $info) use ($services) {
                    $validator = new InputValidator(
                        \func_get_args(),
                        $services->get('container')->get('validator'),
                        $services->get('validatorFactory'),
                        [
                            'address' => [
                                'cascade' => [
                                    'isCollection' => false,
                                    'referenceType' => $services->getType('Address'),
                                ],
                            ],
                            'birthdate' => [
                                'cascade' => [
                                    'groups' => ['group2'],
                                    'isCollection' => false,
                                    'referenceType' => $services->getType('Birthdate'),
                                ],
                            ],
                        ],
                    );
                    return $services->mutation("mutation_mock", $args, $validator);
                },
                'args' => [
                    [
                        'name' => 'groups',
                        'type' => Type::nonNull(Type::listOf(Type::string())),
                    ],
                    [
                        'name' => 'address',
                        'type' => fn() => Type::nonNull($services->getType('Address')),
                    ],
                    [
                        'name' => 'birthdate',
                        'type' => fn() => $services->getType('Birthdate'),
                    ],
                ],
            ],
        ],
    ];

    parent::__construct($configProcessor->process($config));
}
```

#### After

Now validation rules are generated each in the corresponding entry under the `args` key and the resolver closure doesn't contain much logic:

```php
public function __construct(ConfigProcessor $configProcessor, GraphQLServices $services)
{
    $config = [
        'name' => self::NAME,
        'fields' => fn() => [
            'linkedConstraintsValidation' => [
                'type' => Type::boolean(),
                'resolve' => function ($value, $args, $context, $info) use ($services) {
                    $validator = $services->createInputValidator(func_get_args());
                    return $services->mutation("mutation_mock", $args, $validator);
                },
                'args' => [
                    [
                        'name' => 'string1',
                        'type' => Type::nonNull(Type::string()),
                        'validation' => [
                            'link' => ['App\Models\User', 'string1', 'property'],
                        ],
                    ],
                    [
                        'name' => 'string2',
                        'type' => Type::nonNull(Type::string()),
                        'validation' => [
                            'link' => ['App\Models\User', 'string2', 'getter'],
                        ],
                    ],
                    [
                        'name' => 'string3',
                        'type' => Type::nonNull(Type::string()),
                        'validation' => [
                            'link' => ['App\Models\User', 'string3', 'member'],
                        ],
                    ],
                ],
            ],
            'cascadeValidationWithGroups' => [
                'type' => Type::boolean(),
                'resolve' => function ($value, $args, $context, $info) use ($services) {
                    $validator = $services->createInputValidator(func_get_args());
                    return $services->mutation("mutation_mock", $args, $validator);
                },
                'args' => [
                    [
                        'name' => 'groups',
                        'type' => Type::nonNull(Type::listOf(Type::string())),
                    ],
                    [
                        'name' => 'address',
                        'type' => fn() => Type::nonNull($services->getType('Address')),
                        'validation' => InputValidator::CASCADE,
                    ],
                    [
                        'name' => 'birthdate',
                        'type' => fn() => $services->getType('Birthdate'),
                        'validation' => [
                            InputValidator::CASCADE => ['group2'],
                        ],
                    ],
                ],
            ],
        ],
    ];

    parent::__construct($configProcessor->process($config));
}
```

### Additionally

- The `'cascade'` literal is replaced with the constant `InputValidator::CASCADE`
- The creation of the `InputValidator` instances isn't hardcoded inside the resolver closure, but fetched from a service:
   **Before**:
   ```php
    $validator = new InputValidator(
        \func_get_args(),
        $services->get('container')->get('validator'),
        $services->get('validatorFactory'),
        [ /* property constraints */ ],
        [ /* class constraints */]
    );
   ```
  **After**:
  ```php
  $validator = $services->createInputValidator(func_get_args());
  ```
- The keys `isCollection` and `referenceType` are removed as unnecessary, because this information is read directly from the type classes.

### CODE REVIEW

**Review by Vincz — APPROVED** [review]

File: src/Definition/GraphQLServices.php [hunk 1]
```diff
@@ -81,4 +82,12 @@ public function getType(string $typeName): ?Type
     {
         return $this->types->resolve($typeName);
     }
+
+    /**
+     * Creates an instance of InputValidator
+     */
+    public function createInputValidator(array $resolverArgs): InputValidator
```
Comment (mcg-web) [inline src/Definition/GraphQLServices.php:89]: Can we decompose resolve args to have all types definitions?

File: src/Definition/GraphQLServices.php [hunk 2]
```diff
@@ -81,4 +82,12 @@ public function getType(string $typeName): ?Type
...
+    {
+        return $this->services['input_validator_factory']->create($resolverArgs);
```
Comment (mcg-web) [inline src/Definition/GraphQLServices.php:91]: Also for this method args or simply create an object ResolverArgs because we loose al types hints with array.

File: src/DependencyInjection/Compiler/ConfigParserPass.php [hunk 3]
```diff
@@ -177,6 +177,7 @@ private function mappingConfig(array $config, ContainerBuilder $container): arra
 
         // app only config files (yml or xml or graphql)
         if ($mappingConfig['auto_discover']['root_dir'] && $container->hasParameter('kernel.root_dir')) {
```
Comment (mcg-web) [inline src/DependencyInjection/Compiler/ConfigParserPass.php:179]: Why this change in this PR? Please open a dedicated PR for this.

File: src/DependencyInjection/Compiler/ConfigParserPass.php [hunk 4]
```diff
@@ -177,6 +177,7 @@ private function mappingConfig(array $config, ContainerBuilder $container): arra
...
+            // @phpstan-ignore-next-line
             $typesMappings[] = ['dir' => $container->getParameter('kernel.root_dir').'/config/graphql', 'types' => null];
         }
         if ($mappingConfig['auto_discover']['bundles']) {
```
Comment (mcg-web) [inline src/DependencyInjection/Compiler/ConfigParserPass.php:183]: Why this change in this PR? Please open a dedicated PR for this.

File: src/DependencyInjection/Compiler/ConfigParserPass.php [hunk 5]
```diff
@@ -212,10 +213,12 @@ function (array $typeMapping) use ($container) {
...
         $typesMappings = [];
+
+        /** @var array<string, class-string> $bundles */
         $bundles = $container->getParameter('kernel.bundles');
 
         // auto detect from bundle
-        foreach ($bundles as $name => $class) {
+        foreach ($bundles as $class) {
             // skip this bundle
             if (OverblogGraphQLBundle::class === $class) {
```
Comment (mcg-web) [inline src/DependencyInjection/Compiler/ConfigParserPass.php:223]: Why this change in this PR? Please open a dedicated PR for this.

File: src/Generator/TypeBuilder.php [hunk 6]
```diff
@@ -422,16 +415,16 @@ protected function buildScalarCallback($callback, string $fieldName)
...
+     *          return $services->mutation("create_post", $validator]);
      *      }
      *
      * Render example (with validation, but errors are injected into the user-defined resolver):
      * {@link https://github.com/overblog/GraphQLBundle/blob/master/docs/validation/index.md#injecting-errors}
      *
      *      function ($value, $args, $context, $info) use ($services) {
      *          $errors = new ResolveErrors();
-     *          $validator = {@see buildValidatorInstance}
+     *          $validator = $services->createInputValidator(func_get_args());
```
Comment (mcg-web) [inline src/Generator/TypeBuilder.php:427]: Spread arguments to use explicit type declaration instead of array

File: src/Generator/TypeBuilder.php [hunk 7]
```diff
@@ -441,36 +434,36 @@ protected function buildScalarCallback($callback, string $fieldName)
...
+                $injectErrors = EL::expressionContainsVar('errors', $resolve);
 
-            $injectValidator = EL::expressionContainsVar('validator', $resolve);
+                if ($injectErrors) {
+                    $closure->append('$errors = ', Instance::new(ResolveErrors::class));
+                }
 
-            if (null !== $validationConfig) {
-                $closure->append('$validator = ', $this->buildValidatorInstance($validationConfig));
+                $closure->append('$validator = ', "$this->gqlServices->createInputValidator(func_get_args())");
```
Comment (mcg-web) [inline src/Generator/TypeBuilder.php:461]: Spread args

File: tests/Functional/App/config/cascadeOnScalars/mapping/Mutation.types.yml [hunk 8]
```diff
@@ -4,7 +4,7 @@ Mutation:
         fields:
             cascadeOnScalar:
                 type: Boolean
-                resolve: '@=mut("mutation_mock", [args, validator])'
+                resolve: '@=mut("mutation_mock", args, validator)'
```
Comment (mcg-web) [inline tests/Functional/App/config/cascadeOnScalars/mapping/Mutation.types.yml:7]: How this could work since the change of the signature? 🤔

**Review by mcg-web — CHANGES_REQUESTED** [review]
Also can you provide an example of generation without no validation present please?

**Comment by murtukov** [comment]
> Also can you provide an example of generation without no validation present please?

If no validation is present, then types are generated as usual, so there is nothing to provide

File: src/DependencyInjection/Compiler/ConfigParserPass.php (hunk 4 above)
Comment (murtukov) [reply src/DependencyInjection/Compiler/ConfigParserPass.php:183]: https://github.com/overblog/GraphQLBundle/pull/816

File: src/DependencyInjection/Compiler/ConfigParserPass.php (hunk 5 above)
Comment (murtukov) [reply src/DependencyInjection/Compiler/ConfigParserPass.php:223]: https://github.com/overblog/GraphQLBundle/pull/816

File: src/DependencyInjection/Compiler/ConfigParserPass.php (hunk 3 above)
Comment (murtukov) [reply src/DependencyInjection/Compiler/ConfigParserPass.php:179]: https://github.com/overblog/GraphQLBundle/pull/816

File: tests/Functional/App/config/cascadeOnScalars/mapping/Mutation.types.yml (hunk 8 above)
Comment (murtukov) [reply tests/Functional/App/config/cascadeOnScalars/mapping/Mutation.types.yml:7]: Forgot to change the alias from `mut` to `m` in the service definition.

Comment (murtukov) [reply tests/Functional/App/config/cascadeOnScalars/mapping/Mutation.types.yml:7]: Fixed here https://github.com/overblog/GraphQLBundle/pull/817

File: src/Definition/GraphQLServices.php (hunk 1 above)
Comment (murtukov) [reply src/Definition/GraphQLServices.php]: done

**Review by mcg-web — APPROVED** [review]
